### ISSUE TITLE
Training QA model with WikiQA and TrecQA

### ISSUE THREAD

**Issue by yumoxu** [castorini/birch#28]
Hi,

Thanks again for your nice work!

I am quite interested in the QA model which was trained on the data described in your ArXiv paper as follows,

> the union of the TrecQA (Yao et al., 2013) and WikiQA (Yang et al., 2015) datasets.

Since I am now also trying to train a similar model and have several minor questions, I would be really appreciative if you could kindly clarify them for me:

1. Did you use the union of the training sets from TrecQA and WikiQA as the training set, the union of their development sets for development? What about test sets, e.g., were they not used? Or, you integrated all the samples from TrecQA and WikiQA, the split it into train/dev manually?
2. In terms of TrecQA, did you use TRAIN (which contains only 94 questions) or TRAIN-ALL (which contains 1229 questions)?
3. In terms of WikiQA, did you truncate answer sentences to 40 tokens as one of the preprocessing steps (as introduced here: https://github.com/castorini/data/tree/master/WikiQA)?
4. I found the following default configurations in args.py; were they also used for training the QA model?
    - epochs: 3
    - learning rate: 3e-6
    - batch_size: 32
    - warmup_proportion: 0.1

Best,
Yumo

**Comment by Victor0118** [comment]
HI @yumoxu Thanks for your interest.

1. We union training set of two datasets as the training set. The same for dev set. We don't use the test set because our target is newswire document retrieval.
2. We use TRAIN-ALL 
3. We use the data here https://git.uwaterloo.ca/jimmylin/Castor-data/tree/master/datasets/WikiQA and there is no truncation step
4. @zeynepakkalyoncu knows the answer

**Comment by zeynepakkalyoncu** [comment]
@yumoxu Thank you for your interest! Regarding your last question: yes, we used the same configuration except batch size is 16 (will update `args.py` accordingly).

**Comment by yumoxu** [comment]
Hi, @Victor0118 and @zeynepakkalyoncu,

Thank you so much for your kind clarification. 

I loaded the released QA model and noticed that it had the following performance saved in the file,

> MAP: 0.8225, MRR: 0.8393, P30: 0.0583

These results look very impressive to me! If my understanding is right, these are the results on the **whole** union dev set of WikiQA and TrecQA (whose size is 2274), instead of being measured on the two dev sets separately and then averaged (can you confirm this?) 

I tried to retrain the model and replicate this performance but however failed; the best I could achieve so far is as follows,

> MAP: 0.7548, MRR: 0.7825, P30: 0.0583

These results were produced at the end of the second training epoch. Since there are fairly large gaps in terms of MAP and MRR, there might be some key points I missed and I would be really grateful if you could offer me some suggestions.

After carefully mapping the original training and dev files from the two datasets (i.e., a.toks, b.toks, sim.txt, id.txt, and qrel) into one single training csv and one single dev csv, I did the following modifications to your code,

1. In **model.data.py**, I added the support for loading instances from QA collection in **DataGenerator** as shown here:
https://github.com/yumoxu/birch/blob/9ba4484aecba69943ce90938fc249c43a3b11997/src/model/data.py#L36-L39;

2. In **model.data.py**, since the qids in TrecQA dev is in the format of '1.3', '2.4', and therefore are tricky to be converted to int tensor (or float then back to str), in the function **load_data**, I included qids and docids in batches without the conversion,
https://github.com/yumoxu/birch/blob/9ba4484aecba69943ce90938fc249c43a3b11997/src/model/data.py#L96-L99 
Personally I don't think it affects anything since qids and docids are only for being written into prediction file for trec_eval; please correct me if it is wrong.

3. The size of the union training set is 62809 and thus I modified the 41579 in the following line to 62809,
https://github.com/castorini/birch/blob/2dd0401ebb388a1c96f8f3357a064164a5db3f0e/src/model/utils.py#L31 

I also used the following configurations:
- epochs: 3
- learning rate: 3e-6
- batch_size: 16
- warmup_proportion: 0.1
- eval_steps: 1000

Since your codebase is mainly for MB training and I am replicating based on it, I guess there could be some settings specifically for QA training that are not included. If so, it will be really helpful if you could kindly check those for me.

P.S. I noticed that training samples were not shuffled in **DataGenerator**; I think it might potentially hurt the training procedure but even with this you can still achieve such good performance. That's awesome!

**Comment by Victor0118** [comment]
> These results look very impressive to me! If my understanding is right, these are the results on the whole union dev set of WikiQA and TrecQA (whose size is 2274), instead of being measured on the two dev sets separately and then averaged (can you confirm this?)

Yes. You are right.

> I noticed that training samples were not shuffled in DataGenerator; I think it might potentially hurt the training procedure but even with this you can still achieve such good performance. 

I think we did find shuffling can improve the performance, and for the results reported in the paper, we did include data shuffling. @zeynepakkalyoncu Can you check why there is no data shuffling in the current code base? I think we discussed this months ago.

For the experiments on the QA data, I will check it for you soon.

**Comment by Victor0118** [comment]
@yumoxu Seems that I used the raw version of TrecQA. (`raw-dev`) Could you check if it is the same with yours?

Btw, be careful when you merge the IDs from two datasets since there might be overlaps.

**Comment by yumoxu** [comment]
@Victor0118 Thanks for getting back to me!

> I think we did find shuffling can improve the performance, and for the results reported in the paper, we did include data shuffling.

It's good to know shuffle was used. Thanks!

> Seems that I used the raw version of TrecQA

Yes, I can confirm I am using `raw-dev` for TrecQA as well.

> Btw, be careful when you merge the IDs from two datasets since there might be overlaps.

I did notice that there are overlaps of `qids` between the training sets of WikiQA and TrecQA; but in terms of dev sets which is the data used for `trec_evel`, `qids` are distinguishable between the two since TrecQA's qids are like `1.4, 1.5` while WikiQA's `qids` are integers, e.g., `2, 3`. 

Another possible overlap is between `doc_ids`; but I assume it does not matter as long as the `doc_ids` for the same query are distinguishable (**please correct me if I am wrong**). For instance, it's not problematic for the `doc_ids` in WikiQA dev starts from `1` for every query and are thus overlapped.

Would it be possible for you to kindly share with me the union data you used (or the merge script) so I could check if the actual problem is from data or training?

**Comment by Victor0118** [comment]
```
cat ../Castor-data/datasets/TrecQA/train-all/a.toks > ./train/a.toks
cat ../Castor-data/datasets/TrecQA/raw-dev/a.toks >> ./dev/a.toks
cat ../Castor-data/datasets/TrecQA/raw-test/a.toks >> ./test/a.toks

cat ../Castor-data/datasets/TrecQA/train-all/b.toks > ./train/b.toks
cat ../Castor-data/datasets/TrecQA/raw-dev/b.toks >> ./dev/b.toks
cat ../Castor-data/datasets/TrecQA/raw-test/b.toks >> ./test/b.toks

cat ../Castor-data/datasets/TrecQA/train-all/id.txt > ./train/id.txt
cat ../Castor-data/datasets/TrecQA/raw-dev/id.txt >> ./dev/id.txt
cat ../Castor-data/datasets/TrecQA/raw-test/id.txt >> ./test/id.txt

cat ../Castor-data/datasets/TrecQA/train-all/sim.txt > ./train/sim.txt
cat ../Castor-data/datasets/TrecQA/raw-dev/sim.txt >> ./dev/sim.txt
cat ../Castor-data/datasets/TrecQA/raw-test/sim.txt >> ./test/sim.txt


cat ../Castor-data/datasets/WikiQA/train/a.toks >> ./train/a.toks
cat ../Castor-data/datasets/WikiQA/dev/a.toks >> ./dev/a.toks
cat ../Castor-data/datasets/WikiQA/test/a.toks >> ./test/a.toks

cat ../Castor-data/datasets/WikiQA/train/b.toks >> ./train/b.toks
cat ../Castor-data/datasets/WikiQA/dev/b.toks >> ./dev/b.toks
cat ../Castor-data/datasets/WikiQA/test/b.toks >> ./test/b.toks

awk '$1="100"$1' ../Castor-data/datasets/WikiQA/train/id.txt >> ./train/id.txt
awk '$1="200"$1' ../Castor-data/datasets/WikiQA/dev/id.txt >> ./dev/id.txt
awk '$1="300"$1' ../Castor-data/datasets/WikiQA/test/id.txt >> ./test/id.txt

cat ../Castor-data/datasets/WikiQA/train/sim.txt >> ./train/sim.txt
cat ../Castor-data/datasets/WikiQA/dev/sim.txt >> ./dev/sim.txt
cat ../Castor-data/datasets/WikiQA/test/sim.txt >> ./test/sim.txt
```

**Comment by yumoxu** [comment]
@Victor0118 Thanks Victor! You didn't mention `qrel` files but I guess I only need to apply the `200` prefix to the qids in `dev.qrel` of WikiQA also, and then concatenate them with TrecQA? 

I will rebuild my data, do the data shuffling, retrain my model, and then get back to you :)
